homeassistant/components/deluge/const.py

```diff
@@ -1,17 +1,37 @@
 """Constants for the Deluge integration."""
 
+import enum
 import logging
 from typing import Final
 
 CONF_WEB_PORT = "web_port"
 CURRENT_STATUS = "current_status"
-DATA_KEYS = ["upload_rate", "download_rate", "dht_upload_rate", "dht_download_rate"]
 DEFAULT_NAME = "Deluge"
 DEFAULT_RPC_PORT = 58846
 DEFAULT_WEB_PORT = 8112
 DOMAIN: Final = "deluge"
 DOWNLOAD_SPEED = "download_speed"
 
 LOGGER = logging.getLogger(__package__)
-
+PROTOCOL_TRAFFIC_DOWNLOAD_SPEED = "protocol_traffic_download_speed"
+PROTOCOL_TRAFFIC_UPLOAD_SPEED = "protocol_traffic_upload_speed"
 UPLOAD_SPEED = "upload_speed"
+
+
+class DelugeGetSessionStatusKeys(enum.Enum):
```

Comment:
Make it a StrEnum, then each value is still a string and you can avoid the .value

Reply:
so i looked into this, i made the change but deluge-client doesn't like this

it seems that Deluge's XMLRPC is using Bencoding (aka the binary serialization used in torrents), and it is trying to convert the arguments based on the `type()` of the object, and it blows up when it is not a primitive type that it knows how to convert

you can see the code here: https://github.com/JohnDoee/deluge-client/blob/abca22ccb300593790aad6651be583a9b06229a7/deluge_client/rencode.py#L378C1-L379C39

and here it is in the python REPL:

```plaintext
>>> class DelugeKey(enum.Enum):
...  DownloadSpeed="download_rate"
...
>>> x.call("core.get_session_status", [x.value for x in list(DelugeKey)])
{b'download_rate': 947.5}
>>> x.call("core.get_session_status", list(DelugeKey))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\auror\AppData\Roaming\Python\Python312\site-packages\deluge_client\client.py", line 265, in call
    self._send_call(self.deluge_version, self.deluge_protocol_version, method, *args, **kwargs)
  File "C:\Users\auror\AppData\Roaming\Python\Python312\site-packages\deluge_client\client.py", line 165, in _send_call
    req = zlib.compress(dumps(req))
                        ^^^^^^^^^^
  File "C:\Users\auror\AppData\Roaming\Python\Python312\site-packages\deluge_client\rencode.py", line 429, in dumps
    encode_func[type(x)](x, r)
  File "C:\Users\auror\AppData\Roaming\Python\Python312\site-packages\deluge_client\rencode.py", line 379, in encode_list
    encode_func[type(i)](i, r)
  File "C:\Users\auror\AppData\Roaming\Python\Python312\site-packages\deluge_client\rencode.py", line 379, in encode_list
    encode_func[type(i)](i, r)
  File "C:\Users\auror\AppData\Roaming\Python\Python312\site-packages\deluge_client\rencode.py", line 379, in encode_list
    encode_func[type(i)](i, r)
  [Previous line repeated 1 more time]
KeyError: <enum 'DelugeKey'>
>>>
```